Stop warning on .names() access in max_pool2d and max_pool2d_backward

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60059 Stop warning on .names() access in max_pool2d and max_pool2d_backward**

Fixes #60053.

The problem is that `.names()` always triggers the named tensor warning.
To not trigger it, one has to guard it with has_names:
`x.has_names() ? x.names() : DimnameList{}`

This is not the first time this has happened; we should probably
make it so that .names() doesn't raise a warning unless it is actually
populated with names. That's a little tricky to implement so I'm leaving
it for the future.

Test Plan:
- New test, also run `python test/test_nn.py -v -k "max_pool"` and
confirm there are no warnings.

Differential Revision: [D29152737](https://our.internmc.facebook.com/intern/diff/D29152737)

zou3519 committed:
Stop warning on .names() access in max_pool2d and max_pool2d_backward

Fixes #60053.

The problem is that `.names()` always triggers the named tensor warning.
To not trigger it, one has to guard it with has_names:
`x.has_names() ? x.names() : DimnameList{}`

This is not the first time this has happened; we should probably
make it so that .names() doesn't raise a warning unless it is actually
populated with names. That's a little tricky to implement so I'm leaving
it for the future.

Test Plan:
- New test, also run `python test/test_nn.py -v -k "max_pool"` and
confirm there are no warnings.

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29152737).

gchanan commented on test/test_namedtensor.py:
```diff
@@ -296,6 +296,14 @@ def check_tuple_return(op, inputs, expected_names):
             check_tuple_return(F.max_pool2d_with_indices, [named_tensor_2d, [2, 2]], named_tensor_2d.names)
             check_tuple_return(F.max_pool3d_with_indices, [named_tensor_3d, [2, 2, 2]], named_tensor_3d.names)
 
+    def test_max_pooling_without_names_does_not_warn(self):
```
test cuda

gchanan commented:
do we have a check that the names are checked correctly?  I don't see where that would be happening.

zou3519 committed:
Update on "Stop warning on .names() access in max_pool2d and max_pool2d_backward"


Fixes #60053.

The problem is that `.names()` always triggers the named tensor warning.
To not trigger it, one has to guard it with has_names:
`x.has_names() ? x.names() : DimnameList{}`

This is not the first time this has happened; we should probably
make it so that .names() doesn't raise a warning unless it is actually
populated with names. That's a little tricky to implement so I'm leaving
it for the future.

Test Plan:
- New test, also run `python test/test_nn.py -v -k "max_pool"` and
confirm there are no warnings.

Differential Revision: [D29152737](https://our.internmc.facebook.com/intern/diff/D29152737)

[ghstack-poisoned]

zou3519 commented:
> do we have a check that the names are checked correctly? I don't see where that would be happening.

There's no name checking in max_pool IIRC: because max_pool accepts a single tensor, the behavior is that the result tensor should have the same names as the input tensor. (We didn't decide to e.g. enforce that the input to max_pool2d has names N C H W)

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29152737).
